pkg/cluster/agent_customization.go

```diff
@@ -0,0 +1,73 @@
+package cluster
+
+import (
+	v3 "github.com/rancher/rancher/pkg/apis/management.cattle.io/v3"
+	"github.com/rancher/rancher/pkg/settings"
+	corev1 "k8s.io/api/core/v1"
+)
+
+// GetClusterAgentTolerations returns additional tolerations for the cluster agent if they have been user defined. If
+// not, nil is returned.
+func GetClusterAgentTolerations(cluster *v3.Cluster) []corev1.Toleration {
+	if cluster.Spec.ClusterAgentDeploymentCustomization != nil &&
+		cluster.Spec.ClusterAgentDeploymentCustomization.AppendTolerations != nil {
+		return cluster.Spec.ClusterAgentDeploymentCustomization.AppendTolerations
+	}
+
+	return nil
+}
+
+// GetClusterAgentAffinity returns node affinity for the cluster agent if it has been user defined. If not, then the
+// default affinity is returned.
+func GetClusterAgentAffinity(cluster *v3.Cluster) *corev1.Affinity {
+	if cluster.Spec.ClusterAgentDeploymentCustomization != nil &&
+		cluster.Spec.ClusterAgentDeploymentCustomization.OverrideAffinity != nil {
+		return cluster.Spec.ClusterAgentDeploymentCustomization.OverrideAffinity
+	}
+
+	return settings.GetClusterAgentDefaultAffinity()
```

Comment:
using `settings.GetClusterAgentDefaultAffinity()` will only return the hard coded affinity within the rancher source code (`setting.go:242`), and not the value could be set by the user as an env var (`setting.go:133`). Is that intended? 

`settings.ClusterAgentDefaultAffinity.Get()` vs `settings.GetClusterAgentDefaultAffinity`

Reply:
@HarrisonWAffel Yes, that's intended.

Reply:
We should be respecting the new values if user updates default affinity setting, not just hardcoded struct from settings.go. When you use `settings.ClusterAgentDefaultAffinity.Get()`, it returns the right expected value (default OR user). The hardcoded value in setting.go is to initialize the setting a default value on startup. But everywhere else, we'd need to use `Get()`. It's not just values set by env var but also what users set via API/kubectl.

Reply:
@a-blender , I don't believe that's what we want.
If `cluster-agent-default-affinity` setting is changed, then clusters deployed without ClusterAgentDeploymentCustomization / ClusterAgentDeploymentCustomization.OverrideAffinity should get the affinity per value of `cluster-agent-default-affinity`, not the default hardcoded one.
UI will show affinity as per what's in `cluster-agent-default-affinity` so if the user chooses to not override it that's what they should get in the provisioned cluster.
cc: @kinarashah

Reply:
> UI will show affinity as per what's in cluster-agent-default-affinity so if the user chooses to not override it that's what they should get in the provisioned cluster. 

@snasovich I think we are on the same page about which value to use, tho where is UI showing this? My understanding is UI is not exposing this setting.

Reply:
@kinarashah , on the Create / Edit Cluster page that allows the user to override affinity. I'm not sure if it will be shown right away or when some sort of "Override affinity" checkbox is checked - but the UI will be prefilled based on the JSON returned in `cluster-agent-default-affinity` setting and the users can change it as necessary (add/remove rules etc.).

Reply:
@snasovich Gotcha, okay. So that's not to change the setting itself but to help ease the configuration of `OverrideAffinity` for users. To clarify, my expectation is if users don't change anything then UI shouldn't send any value in `OverrideAffinity` instead of always setting the value from the setting..is that correct?

Reply:
@kinarashah , correct. If the user chooses to not override the affinity UI should send `nil` in `ClusterAgentDeploymentCustomization` / `ClusterAgentDeploymentCustomization.OverrideAffinity` which back-end should interpret as using whatever is configured in `cluster-agent-default-affinity`.

Obviously, the user may choose to override affinity and send the value in `ClusterAgentDeploymentCustomization.OverrideAffinity` that matches `cluster-agent-default-affinity` in which case it's no longer actually using the default and whatever change is done to `cluster-agent-default-affinity` will never affect those clusters.